File: src/main/java/com/imyyq/mvvm/binding/viewadapter/scrollview/ViewAdapter.java
package com.imyyq.mvvm.binding.viewadapter.scrollview;

import android.widget.ScrollView;

import androidx.core.widget.NestedScrollView;
import androidx.databinding.BindingAdapter;

import com.imyyq.mvvm.binding.command.BindingCommand;

public final class ViewAdapter {

    @BindingAdapter({"onScrollChangeCommand"})
    public static void onScrollChangeCommand(final NestedScrollView nestedScrollView, final BindingCommand<NestScrollDataWrapper> onScrollChangeCommand) {
        nestedScrollView.setOnScrollChangeListener((NestedScrollView.OnScrollChangeListener) (v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
            if (onScrollChangeCommand != null) {
                onScrollChangeCommand.execute(new NestScrollDataWrapper(scrollX, scrollY, oldScrollX, oldScrollY));
            }
        });
    }

    @BindingAdapter({"onScrollChangeCommand"})
    public static void onScrollChangeCommand(final ScrollView scrollView, final BindingCommand<ScrollDataWrapper> onScrollChangeCommand) {
        scrollView.getViewTreeObserver().addOnScrollChangedListener(() -> {
            if (onScrollChangeCommand != null) {
                onScrollChangeCommand.execute(new ScrollDataWrapper(scrollView.getScrollX(), scrollView.getScrollY()));
            }
        });
    }

    public static class ScrollDataWrapper {
        public float scrollX;
        public float scrollY;

        public ScrollDataWrapper(float scrollX, float scrollY) {
            this.scrollX = scrollX;
            this.scrollY = scrollY;
        }
    }

    public static class NestScrollDataWrapper {
        public int scrollX;
        public int scrollY;
        public int oldScrollX;
        public int oldScrollY;

        public NestScrollDataWrapper(int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
            this.scrollX = scrollX;
            this.scrollY = scrollY;
            this.oldScrollX = oldScrollX;
            this.oldScrollY = oldScrollY;
        }
    }
}

File: src/main/java/com/imyyq/mvvm/binding/viewadapter/recyclerview/ViewAdapter.kt
package com.imyyq.mvvm.binding.viewadapter.recyclerview

import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.RecyclerView
import com.imyyq.mvvm.binding.command.BindingConsumer
import com.imyyq.mvvm.binding.command.BindingFunction

@BindingAdapter("lineManager")
fun setLineManager(
    recyclerView: RecyclerView,
    function: BindingFunction<RecyclerView, RecyclerView.ItemDecoration>
) {
    recyclerView.addItemDecoration(function.apply(recyclerView))
}

@BindingAdapter("layoutManager")
fun setLayoutManager(
    recyclerView: RecyclerView,
    function: BindingFunction<RecyclerView, RecyclerView.LayoutManager>
) {
    recyclerView.layoutManager = function.apply(recyclerView)
}

@BindingAdapter(
    value = ["onScrollChangeCommand", "onScrollStateChangedCommand"],
    requireAll = false
)
fun onScrollChangeCommand(
    recyclerView: RecyclerView,
    onScrollChangeCommand: BindingConsumer<ScrollDataWrapper?>?,
    onScrollStateChangedCommand: BindingConsumer<Int?>?
) {
    recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
        private var state = 0
        private lateinit var wrapper: ScrollDataWrapper
        override fun onScrolled(
            recyclerView: RecyclerView,
            dx: Int,
            dy: Int
        ) {
            super.onScrolled(recyclerView, dx, dy)
            if (onScrollChangeCommand != null) {
                if (!this::wrapper.isInitialized) {
                    wrapper = ScrollDataWrapper()
                }
                wrapper.scrollX = dx.toFloat()
                wrapper.scrollY = dy.toFloat()
                wrapper.state = state
                onScrollChangeCommand.call(wrapper)
            }
        }

        override fun onScrollStateChanged(
            recyclerView: RecyclerView,
            newState: Int
        ) {
            super.onScrollStateChanged(recyclerView, newState)
            state = newState
            onScrollStateChangedCommand?.call(newState)
        }
    })
}

class ScrollDataWrapper {
    var scrollX = 0f
    var scrollY = 0f
    var state = 0
    override fun toString(): String {
        return "ScrollDataWrapper{" +
                "scrollX=" + scrollX +
                ", scrollY=" + scrollY +
                ", state=" + state +
                '}'
    }
}